src/swarm-lang/Swarm/Language/Types.hs

```diff
@@ -429,5 +432,84 @@ pattern UTyCmd ty = UTyConApp TCCmd [ty]
 pattern UTyDelay :: UType -> UType
 pattern UTyDelay ty = UTyConApp TCDelay [ty]
 
+pattern UTyUser :: Var -> [UType] -> UType
+pattern UTyUser v tys = UTyConApp (TCUser v) tys
+
 pattern PolyUnit :: Polytype
 pattern PolyUnit = Forall [] (TyCmd TyUnit)
+
+------------------------------------------------------------
+-- Contexts
+------------------------------------------------------------
+
+-- | A @TCtx@ is a mapping from variables to polytypes.  We generally
+--   get one of these at the end of the type inference process.
+type TCtx = Ctx Polytype
+
+-- | A @UCtx@ is a mapping from variables to polytypes with
+--   unification variables.  We generally have one of these while we
+--   are in the midst of the type inference process.
+type UCtx = Ctx UPolytype
+
+------------------------------------------------------------
+-- Type definitions
+------------------------------------------------------------
+
+data TydefInfo = TydefInfo
+  { _tydefType :: Polytype
+  , _tydefArity :: Arity
+  }
+  deriving (Eq, Show, Generic, Data, FromJSON, ToJSON)
+
+makeLenses ''TydefInfo
+
+-- | A @TDCtx@ is a mapping from user-defined type constructor names
+--   to their definitions and arities/kinds.
+type TDCtx = Ctx TydefInfo
+
+-- | Expand an application "T ty1 ty2 ... tyn" by looking up the
+--   definition of T and substituting ty1 .. tyn for its arguments.
+--
+--   Note that this has already been kind-checked so we know the
+--   number of arguments must match up; we don't worry about what
+--   happens if the lists have different lengths since in theory that
+--   can never happen.
+expandTydef :: Has (Reader TDCtx) sig m => Var -> [UType] -> m UType
+expandTydef userTyCon tys = do
+  mtydefInfo <- Ctx.lookupR userTyCon
+  tdCtx <- ask @TDCtx
+  -- In theory, if everything has kind checked, we should never encounter an undefined
+  -- type constructor here.
+  let tydefInfo = fromMaybe (error $ "Encountered undefined type constructor " ++ into @String userTyCon ++ " in expandTyDef (tdCtx = " ++ show tdCtx ++ ")") mtydefInfo
```

Comment:
Maybe better to use our `Text` helpers like `parens`, then convert to `String` at the end?